readthedocs/search/serializers.py

```diff
@@ -58,11 +59,26 @@ class PageSearchSerializer(serializers.Serializer):
     project = serializers.CharField()
     version = serializers.CharField()
     title = serializers.CharField()
-    link = serializers.SerializerMethodField()
+    path = serializers.SerializerMethodField()
+    domain = serializers.SerializerMethodField()
     highlights = PageHighlightSerializer(source='meta.highlight', default=dict)
     blocks = serializers.SerializerMethodField()
 
-    def get_link(self, obj):
+    def get_domain(self, obj):
```

Comment:
This definitely seems like it's adding a bunch of queries on both of these functions to do the full `resolve` and then ignore parts of it (eg. we don't care about subprojects for the domain). Is there a reason not to just call the resolver `resolve_path` and `resolve_domain` directly here?

Reply:
The result from `_get_full_path` is cached, and we already pass `project_data` into the context, so this won't generate any extra queries.

Reply:
https://github.com/readthedocs/readthedocs.org/blob/7274123c2c970e70d79542aba969c38864d9a376/readthedocs/search/api.py#L379-L382

Reply:
Calling resolve_path and resolve_domain here will generate extra queries.

Reply:
It seems like `projects_data` is only used in this one place, so I don't understand why we're caching it prior to calling this code? Seems like we could just remove all the pre-setting and only set it here when we actually use it?

Reply:
Basically we get all the data in one query more or less https://github.com/readthedocs/readthedocs.org/blob/7274123c2c970e70d79542aba969c38864d9a376/readthedocs/search/api.py#L259-L296

Reply:
Yea, but this line does a ton of queries, no? https://github.com/readthedocs/readthedocs.org/blob/7274123c2c970e70d79542aba969c38864d9a376/readthedocs/search/api.py#L284

Reply:
The serializer only knows about one object, not about all of them. But the caller of this class has the list of all objects that the serializer is going to use, so it can retrieve all the data in one query.

Reply:
Sure, but you're doing queries for every Domain for every subproject with this approach, instead of querying the doctype for every Version, which will lead to the same number of queries?

Reply:
And we already access `obj.version` here, so that wouldn't be any additional queries to just get the doctype from there.

Reply:
Ok, I see what you mean, yeah, that can be optimized to query the domain only once.

Reply:
I don't think we need to worry too much about making this super efficient -- I'm actually saying we should make the code simpler rather than try and make it super fast. We don't do _that_ many searches, so having simple code is probably better. I guess it might matter for projects with a lot of subprojects.

Reply:
I also think we need to think a bit more deeply about how to make this stuff faster at the resolver level, rather than trying to optimize specific areas. We've done this a few times, and really the solution should be "calling the resolver is always fast"